File: leetcode/medium/0098ValidateBST/golang/valid.go
package golang

// TreeNode data class
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBSTIterative(root *TreeNode) bool {
	MIN, MAX := -1<<63, 1<<63-1
	if root == nil {
		return true
	}
	stack := []*TreeNode{}
	upper := []int{}
	lower := []int{}

	stack = append(stack, root)
	upper = append(upper, MAX)
	lower = append(lower, MIN)

	for len(stack) != 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		left := lower[len(lower)-1]
		lower = lower[:len(lower)-1]

		right := upper[len(upper)-1]
		upper = upper[:len(upper)-1]

		if current.Val > right || current.Val < left {
			return false
		}
		if current.Right != nil {
			if current.Right.Val <= current.Val {
				return false
			}
			stack = append(stack, current.Right)
			lower = append(lower, current.Val+1)
			upper = append(upper, right)
		}
		if current.Left != nil {
			if current.Left.Val >= current.Val {
				return false
			}
			stack = append(stack, current.Left)
			lower = append(lower, left)
			upper = append(upper, current.Val-1)
		}
	}
	return true

}

func isValidBST(root *TreeNode) bool {
	MIN, MAX := -1<<63, 1<<63-1
	return recursiveCheck(root, MIN, MAX)
}
func recursiveCheck(node *TreeNode, min int, max int) bool {
	if node == nil {
		return true
	}
	if node.Val < min || node.Val > max {
		return false
	}
	if node.Left != nil && node.Left.Val >= node.Val {
		return false
	}
	if node.Right != nil && node.Right.Val <= node.Val {
		return false
	}

	return recursiveCheck(node.Left, min, node.Val-1) && recursiveCheck(node.Right, node.Val+1, max)
}

File: leetcode/medium/0098ValidateBST/valid.kt
data class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}


class IterativeSolution {
    fun isValidBST(root: TreeNode?): Boolean {
        val intMax = 2147483647
        val intMin = -2147483648
        if (root == null) {
            return true
        }
        
        val stack = Stack<TreeNode>()
        val upper = Stack<Int>()
        val lower = Stack<Int>()
        
        stack.push(root)
        upper.push(intMax)
        lower.push(intMin)
        
        while (stack.isNotEmpty()) {
            val current = stack.pop()
            val left = lower.pop()
            val right = upper.pop()
            
            if (current.`val` > right || current.`val` < left) {
			    return false
		    }
		    if (current.right != null ){
			    if (current.right.`val` <= current.`val`) {
				    return false
			    }
			    stack.push(current.right)
			    lower.push(current.`val` + 1)
			    upper.push(right)
		    }
		    if (current.left != null) {
			    if (current.left.`val` >= current.`val`) {
				    return false
			    }
			    stack.push(current.left)
			    lower.push(left)
			    upper.push(current.`val` - 1)
		    }
        }
        return true
    }
    
}

class RecursiveSolution {
    fun isValidBST(root: TreeNode?): Boolean {
        return recursiveHelper(root, null, null)
    }
    fun recursiveHelper(node: TreeNode?, lower: Int?, upper: Int?): Boolean{
        if(node == null){
            return true
        }
        var value = node.`val`
        if (lower != null && value <= lower) return false
        if (upper != null && value >= upper) return false
        
        if (! recursiveHelper(node.right, value, upper)) return false
        if (! recursiveHelper(node.left, lower, value)) return false
        
        return true
    }
}